upup/pkg/fi/cloudup/awstasks/network_load_balancer.go

```diff
@@ -354,7 +352,16 @@ func (e *NetworkLoadBalancer) Find(c *fi.Context) (*NetworkLoadBalancer, error)
 	}
 
 	for _, az := range lb.AvailabilityZones {
-		actual.Subnets = append(actual.Subnets, &Subnet{ID: az.SubnetId})
+		sm := &SubnetMapping{
+			Subnet: &Subnet{ID: az.SubnetId},
+		}
+		for _, a := range az.LoadBalancerAddresses {
+			if a.PrivateIPv4Address != nil {
+				sm.PrivateIPv4Address = a.PrivateIPv4Address
```

Comment:
In case someone does not specify `privateIPv4Address` when initially deploying the NLB, AWS will enable automatic assignment of IPs from the free IP pool of the subnet. These automatically assigned IPs do not appear in `LoadBalancerAddresses` (you can only find them when looking at the network interfaces of the NLB). Also, changing the IP assignment in any way is not possible afterwards.

So I assume the current behaviour is the correct one. The only thing an admin can modify after initial creation of the NLB is assigning new subnets to previously unassigned AZs (each AZ can have only one subnet). In that case, kops would detect that there is a new SubnetMapping for the NLB and then try to remove that one, which will of course fail. We can't really avoid this situation, because otherwise a removal of a subnet in the spec would be ignored as well as kops can't know the difference.

Reply:
ah I didn't realize the automatically assigned IPs aren't in this field. In that case I agree the current behavior should work.